### ISSUE TITLE
spam filtering

### ISSUE THREAD

**Issue by posativ** [posativ/isso#11]
It's surprisingly difficult to find a usable general spam filter software. That's currently available:
- [DSPAM](http://dspam.sourceforge.net/) – requires a daemon, dafuq?
- [CRM114](http://crm114.sourceforge.net/) – bloated, but there's a stripped down, abandoned [ANSI C implementation](https://github.com/pmundkur/libcrm114) with Python support, but doesn't compile on my system.
- [bayes classifier](https://pypi.python.org/pypi?%3Aaction=search&term=bayes&submit=search) – most are > 4 years old, don't have a test suite etc.

Probably DIY: http://crm114.sourceforge.net/docs/classify_details.txt

**Comment by castarco** [comment]
Hi, I'm interested on contributing spam filtering. Captcha systems aren't enough since there are paid people doing manual spam, not only bots (personally, in my old blog I received a lot of spam, sometimes coming from real humans).

About the spam filtering system, we can use "support vector machines" instead of bayesian filters, which are relatively efficient after the training phase, so DOS attacks are improbable to be successful.

**Comment by fluffy-critter** [comment]
Lately my site (running isso) has started to receive quite the onslaught of spam. I don't know how much is human-posted and how much is from javascript-aware automation stuff but either way, some of them are even gloating about "easiest captcha ever" on their comment, as if their comment is going to be seen by the public or would be indexed by a search engine. (Also why the hell haven't bots figured out that everyone has used `rel="nofollow"` for like 20 years?)

Anyway. Yeah. A plugin system would be great. I"m willing to spend some time working on one.

**Comment by onaralili** [comment]
> External services can be implemented if they are not required for Isso (Akismet is a US service). Nevertheless, spam filtering should by default not rely on any third-party provider (as it defeats the purpose of self-hosting).

Would be great to have a plugin system to integrate with third-parties. As an alternative to Akismet, there is OOPSpam which is GDPR complaint.

**Comment by ix5** [comment]
> Anyway. Yeah. A plugin system would be great. I"m willing to spend some time working on one.

@fluffy-critter did you eventually come up with something?

**Comment by fluffy-critter** [comment]
I haven't had time/energy to work on anything, unfortunately.

**Comment by taoeffect** [comment]
Hey guys: consider PoW as a simpler means of spam filtering:

- https://git.sequentialread.com/forest/pow-captcha
- https://mcaptcha.org/

**Comment by ix5** [comment]
> Hey guys: consider PoW as a simpler means of spam filtering:
> 
>     * https://git.sequentialread.com/forest/pow-captcha
> 
>     * https://mcaptcha.org/

Those two look interesting, but heads up, they require modern browsers and wasm support.

**Comment by fluffy-critter** [comment]
Also I'm not sure what problem that actually solves, beyond making sure someone's idle on a page for a certain amount of time before they submit. Most of the spam I get appears to be submitted by humans who are paid money to defeat CAPTCHAs, as has been the case with most comment spam for at least the past decade.

**Comment by taoeffect** [comment]
It would cut down on some spam, probably not all. You could increase the difficulty most likely in the settings, etc. It's simple to setup. There are tradeoffs with everything. If you really wanted to prevent all spam (and also some legitimate comments), you could charge micropayments over the lightning network. :P

**Comment by ForestJohnson** [comment]
Hi, I just heard about this project and it looks nice!   In the past I made my own project that was similar and I created the `git.sequentialread.com/forest/pow-captcha` for it.  If you would like to try it out, it's hosted here:  

https://sequentialread.com/now-with-comments/#sqr-comment-container

I have also seen spam from humans.  There were some SEO spammers trying to register accounts on our gitea server and post links to their clients businesses.   We were not able to stop them until we implemented a required invite token for registration :( 

To be honest I'm not sure what to do about that kind of spam besides putting the comments into a moderation queue and having someone look at them.   

My "pow-captcha" is not actually a captcha at all, I think it's just a bot deterrent.  Unfortunately its also a deterrent for people who have extremely old  computers  or who run customized browsers with anti-fingerprinting  or new features disabled.   I use it as bot deterrent for other things and I have seen my friends blocked by it because the browser would not allow WebWorkers etc :(  

I think unfortunately spam is always going to be impossible to stop.  Maybe for the next version of my site I will try out isso for comments but skip the moderation queue if the browser was able to solve the captcha. 

> making sure someone's idle on a page for a certain amount of time before they submit.

The animated gif on the ReadMe is sort of intentionally slowed down / it was for a higher difficulty setting than the one I use on my site.  I think you would have to pull out a cell phone from 8 years ago to see it go that slow on my site. On a new computer its so fast that it can't even render the progress bar before its done.

**Comment by fluffy-critter** [comment]
Yeah there's no way to automatically get rid of all spam, but it's still nice to have a means of being able to classify things to apply different moderation policies to them, and possibly be able to specifically whitelist known-good posters so their comments go up immediately.
